Added gravity to Command

JeffAlyanak opened the pull request:
Added the gravity option to command. This is a very useful option for use with crop, extent, etc. 

It's possible this may need to be updated or altered for different versions of imagemagick if the valid parameters differ, but currently it conforms with the docs provided here: https://www.imagemagick.org/script/command-line-options.php#gravity

JeffAlyanak committed:
Added gravity to Command

Added gravity option to command, reference. Added a reference class for gravity which can validate the gravity parameter provided. It's possible this may need to be updated or altered for different versions of imagemagick, but currently conforms with the docs provided (https://www.imagemagick.org/script/command-line-options.php#gravity).

JeffAlyanak committed:
Fixed exception strings

Exception strings were not calling statics correctly.

JeffAlyanak committed:
Added test for Gravity ref class.

JeffAlyanak committed:
Added test of gravity command.

Pierstoval commented on ReferenceClasses/Gravity.php:
```diff
@@ -0,0 +1,55 @@
+
+    public function validate(): string
+    {
+        if ( !in_array($this->value, self::$validGravity) ) {
```
```suggestion
        if (!in_array($this->value, self::$validGravity, true)) {
```
Strict check + CS fix

JeffAlyanak replied:
Resolved in 6d32748.

Pierstoval commented on Tests/CommandTest.php:
```diff
@@ -140,6 +140,31 @@ public function testColorspaceImage(): void
+        $response = $command
+            ->convert($imageToResize)
+            ->gravity('Center')
+			->extent('100x100')
```
```suggestion
	      		->extent('100x100')
```

JeffAlyanak replied:
Resolved in bf170e5.

Pierstoval commented on Tests/CommandTest.php:
```diff
@@ -140,6 +140,31 @@ public function testColorspaceImage(): void
+
+        static::assertFileExists($this->resourcesDir.'/outputs/moon.jpg');
+
+        static::assertFalse($response->hasFailed());
```
```suggestion
```

Unnecessary check since it's already made two instructions before

JeffAlyanak replied:
Resolved in 197908e.

Pierstoval commented on Tests/GravityTest.php:
```diff
@@ -0,0 +1,114 @@
+        try {
+            $validatedGravity = $gravity->validate();
+        } catch (\Exception $e) {
+            static::fail($e->getMessage());
```
There's no need to force the test to fail. PHPUnit will output an `E` character instead of `F` if an exception is thrown and that'll be okay.

JeffAlyanak replied:
Resolved in 31a8eeb.

Pierstoval commented on Tests/GravityTest.php:
```diff
@@ -0,0 +1,114 @@
+        static::assertNotEmpty($validatedGravity);
+
+        if ('' === $validatedGravity) {
+            static::markTestSkipped('No gravity to check. ['.$validatedGravity.'] ['.\implode(',', \func_get_args()).']');
```
Useless check, since `assertNotEmpty()` will throw an exception if `$validatedGravity` is equal to `''`

JeffAlyanak replied:
Resolved by 31a8eeb.

Pierstoval commented on Tests/GravityTest.php:
```diff
@@ -0,0 +1,114 @@
+        $message = null;
+
+        try {
+            $testGravity->validate();
```
You can use `$this->expectException(\InvalidArgumentException::class);` and `$this->expectExceptionMessage('...');` instead

JeffAlyanak replied:
Resolved in commit 1a10dd2.

Pierstoval commented:
I know this option, thanks for contributing! I'm sorry for the amount of requested changes 😄

JeffAlyanak committed:
Strict in_array check.

JeffAlyanak committed:
Replaced hard tabs with soft.

JeffAlyanak committed:
Removed useless & Redundant Tests

JeffAlyanak committed:
Simplified exception tests for Gravity validatation.

In future the testWrongGeometry function in the Tests/GeometryTest.php can also be cleaned up in a similar way.

JeffAlyanak committed:
Removed redundant try/catch.

Leaving this to phpunit to catch.

JeffAlyanak commented:
> I know this option, thanks for contributing! I'm sorry for the amount of requested changes smile

No problem at all, I've put all of those changes in as of 31a8eeb.

Pierstoval commented on Tests/GravityTest.php:
```diff
@@ -0,0 +1,97 @@
+
+        $validatedGravity = $gravity->validate();
+
+        static::assertIsString($validatedGravity);
```
```suggestion
```

Useless check too, since `->validate()`'s return type is `string`

JeffAlyanak replied:
Fair enough, resolved in 8e5d443.

Pierstoval reviewed (changes requested):
Almost finished 👍

JeffAlyanak committed:
Removed redundant assertIsString

Unless the return type of of Gravity class' validate function is changed, this will always be a string.

Pierstoval reviewed (approved):
Thanks @JeffAlyanak 😄

Pierstoval commented:
Just released [3.0.10](https://github.com/Orbitale/ImageMagickPHP/releases/tag/v3.0.10) 🚀
